// serverless-offline-sns/dist/sns-server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var aws_sdk_1 = require("aws-sdk");
var node_fetch_1 = require("node-fetch");
var url_1 = require("url");
var bodyParser = require("body-parser");
var xml = require("xml");
var helpers_1 = require("./helpers");
var SNSServer = /** @class */ (function () {
    function SNSServer(debug, app, region, accountId) {
        this.pluginDebug = debug;
        this.topics = [];
        this.subscriptions = [];
        this.app = app;
        this.region = region;
        this.routes();
        this.accountId = accountId;
    }
    SNSServer.prototype.routes = function () {
        var _this = this;
        this.debug("configuring route");
        this.app.use(bodyParser.json()); // for parsing application/json
        this.app.use(bodyParser.urlencoded({ extended: true })); // for parsing application/x-www-form-urlencoded
        this.app.use(function (req, res, next) {
            res.header("Access-Control-Allow-Origin", "*");
            res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
            next();
        });
        this.app.all("/", function (req, res) {
            _this.debug("hello request");
            _this.debug(JSON.stringify(req.body));
            _this.debug(JSON.stringify(_this.subscriptions));
            if (req.body.Action === "ListSubscriptions") {
                _this.debug("sending: " + xml(_this.listSubscriptions(), { indent: "\t" }));
                res.send(xml(_this.listSubscriptions()));
            }
            else if (req.body.Action === "CreateTopic") {
                res.send(xml(_this.createTopic(req.body.Name)));
            }
            else if (req.body.Action === "Subscribe") {
                res.send(xml(_this.subscribe(req.body.Endpoint, req.body.Protocol, req.body.TopicArn)));
            }
            else if (req.body.Action === "Publish") {
                res.send(xml(_this.publish(req.body.TopicArn, req.body.subject, req.body.Message, req.body.MessageStructure, helpers_1.parseMessageAttributes(req.body))));
            }
            else if (req.body.Action === "Unsubscribe") {
                res.send(xml(_this.unsubscribe(req.body.SubscriptionArn)));
            }
            else {
                res.send(xml({
                    NotImplementedResponse: [
                        helpers_1.createAttr(),
                        helpers_1.createMetadata(),
                    ],
                }));
            }
            _this.debug(JSON.stringify(_this.subscriptions));
        });
    };
    SNSServer.prototype.listSubscriptions = function () {
        this.debug(this.subscriptions.map(function (sub) {
            return {
                member: [sub],
            };
        }));
        return {
            ListSubscriptionsResponse: [
                helpers_1.createAttr(),
                helpers_1.createMetadata(),
                {
                    ListSubscriptionsResult: [{
                            Subscriptions: this.subscriptions.map(function (sub) {
                                return {
                                    member: helpers_1.arrayify({
                                        Endpoint: sub.Endpoint,
                                        TopicArn: sub.TopicArn,
                                        Owner: sub.Owner,
                                        Protocol: sub.Protocol,
                                        SubscriptionArn: sub.SubscriptionArn,
                                    }),
                                };
                            }),
                        }],
                },
            ],
        };
    };
    SNSServer.prototype.unsubscribe = function (arn) {
        this.debug(JSON.stringify(this.subscriptions));
        this.debug("unsubscribing: " + arn);
        this.subscriptions = this.subscriptions.filter(function (sub) { return sub.SubscriptionArn !== arn; });
        return {
            UnsubscribeResponse: [
                helpers_1.createAttr(),
                helpers_1.createMetadata(),
            ],
        };
    };
    SNSServer.prototype.createTopic = function (topicName) {
        var topic = {
            TopicArn: "arn:aws:sns:" + this.region + ":" + this.accountId + ":" + topicName,
        };
        this.topics.push(topic);
        return {
            CreateTopicResponse: [
                helpers_1.createAttr(),
                helpers_1.createMetadata(),
                {
                    CreateTopicResult: [
                        {
                            TopicArn: topic.TopicArn,
                        },
                    ],
                },
            ],
        };
    };
    SNSServer.prototype.subscribe = function (endpoint, protocol, arn) {
        arn = this.convertPsuedoParams(arn);
        var sub = {
            SubscriptionArn: arn + ":" + Math.floor(Math.random() * (1000000 - 1)),
            Protocol: protocol,
            TopicArn: arn,
            Endpoint: endpoint,
            Owner: "",
        };
        this.subscriptions.push(sub);
        return {
            SubscribeResponse: [
                helpers_1.createAttr(),
                helpers_1.createMetadata(),
                {
                    SubscribeResult: [
                        {
                            SubscriptionArn: sub.SubscriptionArn,
                        },
                    ],
                },
            ],
        };
    };
    SNSServer.prototype.publishHttp = function (event, sub) {
        var _this = this;
        return node_fetch_1.default(sub.Endpoint, {
            method: "POST",
            body: event,
            timeout: 0,
            headers: {
                "Content-Type": "application/json",
                "Content-Length": Buffer.byteLength(event),
            },
        }).then(function (res) { return _this.debug(res); });
    };
    SNSServer.prototype.publishSqs = function (event, sub) {
        var subEndpointUrl = new url_1.URL(sub.Endpoint);
        var sqsEndpoint = subEndpointUrl.protocol + "//" + subEndpointUrl.host + "/";
        var sqs = new aws_sdk_1.SQS({ endpoint: sqsEndpoint, region: this.region });
        return sqs.sendMessage({
            QueueUrl: sub.Endpoint,
            MessageBody: event,
        }).promise();
    };
    SNSServer.prototype.publish = function (topicArn, subject, message, messageType, messageAttributes) {
        var _this = this;
        topicArn = this.convertPsuedoParams(topicArn);
        var messageId = helpers_1.createMessageId();
        Promise.all(this.subscriptions.filter(function (sub) { return sub.TopicArn === topicArn; }).map(function (sub) {
            _this.debug("fetching: " + sub.Endpoint);
            var event = JSON.stringify(helpers_1.createSnsEvent(topicArn, sub.SubscriptionArn, subject, message, messageId, messageAttributes));
            _this.debug("event: " + event);
            if (!sub.Protocol) {
                sub.Protocol = "http";
            }
            var protocol = sub.Protocol.toLowerCase();
            if (protocol === "http") {
                return _this.publishHttp(event, sub);
            }
            if (protocol === "sqs") {
                return _this.publishSqs(event, sub);
            }
            throw new Error("Protocol '" + protocol + "' is not supported by serverless-offline-sns");
        }));
        return {
            PublishResponse: [
                helpers_1.createAttr(),
                {
                    PublishResult: [
                        {
                            MessageId: messageId,
                        },
                    ],
                },
                helpers_1.createMetadata(),
            ],
        };
    };
    SNSServer.prototype.convertPsuedoParams = function (topicArn) {
        var awsRegex = /#{AWS::([a-zA-Z]+)}/g;
        return topicArn.replace(awsRegex, this.accountId);
    };
    SNSServer.prototype.debug = function (msg) {
        this.pluginDebug(msg, "server");
    };
    return SNSServer;
}());
exports.SNSServer = SNSServer;
